pkg/manager/internal.go

```diff
@@ -468,6 +607,81 @@ func (cm *controllerManager) Start(ctx context.Context) (err error) {
 		}()
 	}
 
+	// If the manager has been configured with a logical adapter, start it.
+	if cm.logicalProvider != nil {
+		if err := cm.add(RunnableFunc(func(ctx context.Context) error {
+			resync := func() error {
+				clusterList, err := cm.logicalProvider.List()
+				if err != nil {
+					return err
+				}
+				for _, name := range clusterList {
+					if _, err := cm.getLogicalCluster(ctx, name); err != nil {
+						return err
+					}
+				}
+				clusterListSet := sets.New(clusterList...)
+				for name := range cm.logicalClusters {
+					if clusterListSet.Has(name) {
+						continue
+					}
+					if err := cm.removeLogicalCluster(name); err != nil {
+						return err
+					}
+				}
+				cm.engageClusterAwareRunnables()
+				return nil
+			}
+
+			// Always do an initial full resync.
+			if err := resync(); err != nil {
+				return err
+			}
+
+			// Create a watcher and start watching for changes.
+			watcher, err := cm.logicalProvider.Watch()
+			if err != nil {
+				return err
+			}
+			defer func() {
+				go func() {
+					// Drain the watcher result channel to prevent a goroutine leak.
+					for range watcher.ResultChan() {
+					}
+				}()
+				watcher.Stop()
+			}()
+
+			for {
+				select {
+				case <-time.After(10 * time.Minute):
+					if err := resync(); err != nil {
+						return err
+					}
+				case event := <-watcher.ResultChan():
+					switch event.Type {
+					case watch.Added, watch.Modified:
+						if _, err := cm.getLogicalCluster(ctx, event.Name); err != nil {
+							return err
+						}
+						cm.engageClusterAwareRunnables()
```

Comment:
I wonder if we want to optimize this a bit in a follow-up to not always engage for all clusters, but just for the added/modified one.

Might become a problem with hundreds+ of clusters